File: 1392-longest-happy-prefix/1392-longest-happy-prefix.java
class Solution {
    public String longestPrefix(String s) {
        char[] cs = s.toCharArray();
        long l = 0, r = 0, b = 1, mod = 1000000007; int len = 0; int n = s.length();
        for(int i = 0; i < n - 1; ++i) {
            l = (l * 128 + cs[i]) % mod;
            r = (cs[n - i - 1] * b + r) % mod;
            b = b * 128 % mod;
            if(l == r) len = i + 1;
        }
        return s.substring(0, len);
    }
}

File: 1392-longest-happy-prefix/1392-longest-happy-prefix.cs
public class Solution {
    public string LongestPrefix(string s) {
        int mod = 1000000007;
        long l = 0, r = 0, p = 1; int n = s.Length, len = 0;
        for (int i = 0; i < n - 1; ++i)
        {
            l = (l * 128 + s[i]) % mod;
            r = (r + p * s[n - i - 1]) % mod;
            if (l == r) len = i + 1;
            p = p * 128 % mod;
        }
        return s.Substring(0, (int)len);

    }
}